File: solutions/0157. Read N Characters Given Read4/0157.cpp
/**
 * The read4 API is defined in the parent class Reader4.
 *     int read4(char *buf4);
 */

class Solution {
 public:
  /**
   * @param buf Destination buffer
   * @param n   Number of characters to read
   * @return    The number of actual characters read
   */
  int read(char* buf, int n) {
    char* buf4 = new char[4];  // dynamic buffer
    int i = 0;                 // point to buf

    while (true) {
      const int n4 = read4(buf4);  // will return [0..4]
      for (int i4 = 0; i4 < n4 && i < n; ++i4)
        buf[i++] = buf4[i4];
      if (n4 == 0 || i == n)
        return i;
    }
  }
};

File: solutions/0157. Read N Characters Given Read4/0157.py
"""
The read4 API is already defined for you.
    def read4(buf4: List[chr]) -> int:

# Below is an example of how the read4 API can be called.
file = File("abcdefghijk") # File is "abcdefghijk", initially file pointer (fp) points to 'a'
buf4 = [' '] * 4 # Create buffer with enough space to store characters
read4(buf4) # read4 returns 4. Now buf = ['a','b','c','d'], fp points to 'e'
read4(buf4) # read4 returns 4. Now buf = ['e','f','g','h'], fp points to 'i'
read4(buf4) # read4 returns 3. Now buf = ['i','j','k',...], fp points to end of file
"""


class Solution:
  def read(self, buf: List[str], n: int) -> int:
    buf4 = [' '] * 4
    i4 = 0  # buf4's index
    n4 = 0  # buf4's size
    i = 0  # buf's index

    while i < n:
      if i4 == n4:  # all characters in buf4 are consumed
        i4 = 0  # reset buf4's index
        n4 = read4(buf4)  # read 4 (or less) chars from file to buf4
        if n4 == 0:  # reach the EOF
          return i
      buf[i] = buf4[i4]
      i += 1
      i4 += 1

    return i
